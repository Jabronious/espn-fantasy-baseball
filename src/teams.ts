import { BaseClass } from './base-class';
import { ESPNCookiesDto } from './models/classes/espn-cookies.dto';
import { PlayerDto } from './models/classes/player.dto';
import { TeamDto } from './models/classes/team.dto';

export class Teams extends BaseClass {
	constructor(leagueId: number, cookies?: ESPNCookiesDto) {
		super(leagueId, cookies);
	}

	/**
	 * @param {number} teamId - team id for the team you wish to look up
	 * @returns Team data for the given team id
	 */
	async getDetailedTeamData(teamId: number): Promise<TeamDto> {
		const params = { view: 'mTeam' };
		const response = await this.fantasyRequests.get(`/teams/${teamId}`, {}, params).catch((e) => {
			throw new Error(e);
		});

		return response.data;
	}

	/**
	 * @param {number} teamId - team id for the team you wish to get the roster for
	 * @returns Team roster data for the given team id
	 */
	async getTeamRoster(teamId: number): Promise<PlayerDto[]> {
		const params = { view: 'mRoster' };
		const response = await this.fantasyRequests.get(`/teams/${teamId}`, {}, params).catch((e) => {
			throw new Error(e);
		});

		return this.getPlayersFromEspnRoster(
			response.data.roster as { entries: [{ playerPoolEntry: { player: PlayerDto } }] }
		);
	}

	/**
	 * @returns All team roster data for the league
	 */
	async getAllTeamsRoster(): Promise<{ [key: string]: PlayerDto[] }> {
		const params = { view: 'mRoster' };
		const response = await this.fantasyRequests.get('', {}, params).catch((e) => {
			throw new Error(e);
		});

		const teamData = (
			await this.fantasyRequests.get('', {}, { view: 'mTeam' }).catch((e) => {
				throw new Error(e);
			})
		).data.teams;

		const teamRosters: { [key: string]: PlayerDto[] } = {};
		teamData.forEach((team: { nickname: string; id: number }) => {
			teamRosters[team.nickname] = this.getPlayersFromEspnRoster(
				response.data.teams.find((rosterTeam: { id: number }) => {
					return rosterTeam.id === team.id;
				}).roster
			);
		});
		return teamRosters;
	}

	private getPlayersFromEspnRoster(roster: { entries: [{ playerPoolEntry: { player: PlayerDto } }] }): PlayerDto[] {
		return roster.entries.map((entry: { playerPoolEntry: { player: PlayerDto } }) => {
			return entry.playerPoolEntry.player;
		});
	}
}
