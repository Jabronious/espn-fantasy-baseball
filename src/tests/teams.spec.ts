import axios from 'axios';
import { Teams } from '../teams';
import { ESPNCookiesDto } from '../models/classes/espn-cookies.dto';
import { TeamDto } from '../models/classes/team.dto';
import { PlayerDto } from '../models/classes/player.dto';

const roster = { entries: [{ playerPoolEntry: { player: { fullName: 'Jose Ramirez' } } }] };
describe('Teams', () => {
	const cookies: ESPNCookiesDto = {
		espn_s2: '2B7uOaY7pTx8neDj061z24T3PdDPn6a29EHEMEfelzI6HLT43oy1P%%2BUTkzYAsk%2BkxgrK%',
		swid: '{A11A1111-1AAA-111A-1A11-111AAA111A1A}',
	};
	const teams = new Teams(1, cookies);

	it('should be defined', () => {
		expect(teams).toBeDefined();
	});

	describe('getTeamInfoFromSWID', () => {
		const teamData: Partial<TeamDto>[] = [
			{
				owners: [cookies.swid],
			},
		];
		it('should get team info', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({ data: teamData });

			expect(await teams.getTeamInfoFromSWID()).toMatchSnapshot();
		});

		it('should throw a team not found error', async () => {
			const teamData: Partial<TeamDto>[] = [
				{
					owners: ['12345-12345-1235'],
				},
			];
			jest.spyOn(axios, 'request').mockResolvedValueOnce({ data: teamData });

			await expect(teams.getTeamInfoFromSWID()).rejects.toMatchSnapshot();
		});

		it('should catch errors and throw them', async () => {
			jest.spyOn(axios, 'request').mockRejectedValueOnce('error');

			await expect(teams.getTeamInfoFromSWID()).rejects.toMatchSnapshot();
		});
	});

	describe('getTeamInfoFromSWID', () => {
		const teamData: Partial<TeamDto>[] = [
			{
				owners: [cookies.swid],
				roster: {
					entries: [] as unknown as [{ playerPoolEntry: { player: PlayerDto } }],
				},
			},
		];
		it('should get team info', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({ data: teamData });

			expect(await teams.getTeamRosterFromSWID()).toMatchSnapshot();
		});

		it('should catch errors and throw them', async () => {
			jest.spyOn(axios, 'request').mockRejectedValueOnce('error');

			await expect(teams.getTeamRosterFromSWID()).rejects.toMatchSnapshot();
		});
	});

	describe('getDetailedTeamData', () => {
		it('should get all details for a team', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({ data: { abbrev: 'CDMD' } });

			expect(await teams.getDetailedTeamData(1)).toMatchSnapshot();
		});

		it('should catch errors and throw them', async () => {
			jest.spyOn(axios, 'request').mockRejectedValueOnce('error');

			await expect(teams.getDetailedTeamData(1)).rejects.toMatchSnapshot();
		});
	});

	describe('getTeamRoster', () => {
		it('should get of a teams roster', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({
				data: { roster },
			});

			expect(await teams.getTeamRoster(1)).toMatchSnapshot();
		});

		it('should catch errors and throw them', async () => {
			jest.spyOn(axios, 'request').mockRejectedValueOnce('error');

			await expect(teams.getTeamRoster(1)).rejects.toMatchSnapshot();
		});
	});

	describe('getAllTeamsRoster', () => {
		it('should get all the teams rosters', async () => {
			jest.spyOn(axios, 'request')
				.mockResolvedValueOnce({
					data: {
						teams: [
							{ id: 1, roster },
							{ id: 2, roster },
						],
					},
				})
				.mockResolvedValueOnce({
					data: {
						teams: [
							{ id: 1, name: 'team1' },
							{ id: 2, name: 'team two' },
						],
					},
				});

			expect(await teams.getAllTeamsRoster()).toMatchSnapshot();
		});

		it('should return blank list because there are no teams', async () => {
			jest.spyOn(axios, 'request')
				.mockResolvedValueOnce({
					data: {
						teams: [
							{ id: 1, roster },
							{ id: 2, roster },
						],
					},
				})
				.mockResolvedValueOnce({
					data: {
						teams: [
							{ id: 3, name: 'team1' },
							{ id: 4, name: 'team two' },
						],
					},
				});

			expect(await teams.getAllTeamsRoster()).toMatchSnapshot();
		});

		it('should catch errors and throw them on leagueRosterData', async () => {
			jest.spyOn(axios, 'request').mockRejectedValueOnce('error');

			await expect(teams.getAllTeamsRoster()).rejects.toMatchSnapshot();
		});

		it('should catch errors and throw them on leagueTeamData', async () => {
			jest.spyOn(axios, 'request')
				.mockResolvedValueOnce({
					data: {
						teams: [
							{ id: 1, roster },
							{ id: 2, roster },
						],
					},
				})
				.mockRejectedValueOnce('error');

			await expect(teams.getAllTeamsRoster()).rejects.toMatchSnapshot();
		});
	});
});
