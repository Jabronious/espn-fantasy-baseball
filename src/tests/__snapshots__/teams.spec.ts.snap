// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Teams getAllTeamsRoster should catch errors and throw them 1`] = `[Error: error]`;

exports[`Teams getAllTeamsRoster should get all the teams rosters 1`] = `
Object {
  "team two": Array [
    Object {
      "fullName": "Jose Ramirez",
    },
  ],
  "team1": Array [
    Object {
      "fullName": "Jose Ramirez",
    },
  ],
}
`;

exports[`Teams getDetailedTeamData should catch errors and throw them 1`] = `[Error: error]`;

exports[`Teams getDetailedTeamData should get all details for a team 1`] = `
Object {
  "abbrev": "CDMD",
}
`;

exports[`Teams getTeamRoster should catch errors and throw them 1`] = `[Error: error]`;

exports[`Teams getTeamRoster should get of a teams roster 1`] = `
Array [
  Object {
    "fullName": "Jose Ramirez",
  },
]
`;
